src/main/scala/com/databricks/spark/redshift/Conversions.scala

```diff
@@ -0,0 +1,130 @@
+  val redshiftTimestampFormatWithoutMillis = new SimpleDateFormat(PATTERN_WITHOUT_MILLIS)
+
+  // Redshift may or may not include the fraction component in the UNLOAD data, and there are
+  // apparently not clues about this in the table schema. This format delegates to one of the above
+  // formats based on string length.
+  val redshiftTimestampFormat = new DateFormat() {
+    override def format(date: Date, toAppendTo: StringBuffer, fieldPosition: FieldPosition): StringBuffer = {
+      // Always export with milliseconds, as they can just be zero if not specified
+      redshiftTimestampFormatWithMillis.format(date, toAppendTo, fieldPosition)
+    }
+
+    override def parse(source: String, pos: ParsePosition): Date = {
+      if(source.length < PATTERN_WITH_MILLIS.length) {
+        redshiftTimestampFormatWithoutMillis.parse(source, pos)
+      } else {
+        redshiftTimestampFormatWithMillis.parse(source, pos)
+      }
+    }
+  }
+
+  val redshiftDateFormat = new SimpleDateFormat("yyyy-MM-dd")
+
+  /**
+   * Parse a string exported from a Redshift TIMESTAMP column
+   */
+  def parseTimestamp(s: String): Timestamp = {
+    new Timestamp(redshiftTimestampFormat.parse(s).getTime)
+  }
+
+  /**
+   * Parse a string exported from a Redshift DATE column
+   */
+  def parseDate(s: String): Timestamp = {
+    new Timestamp(redshiftDateFormat.parse(s).getTime)
+  }
+
+  /**
+   * Construct a Row from the given array of strings, retrieved from Redshift UNLOAD.
+   * The schema will be used for type mappings.
+   */
+  def convertRow(schema: StructType, fields: Array[String]): Row = {
+    val converted = fields zip schema map {
+      case (data, field) =>
+        if (data == null || data.isEmpty) null
+        else field.dataType match {
+          case ByteType => data.toByte
+          case BooleanType => RedshiftBooleanParser.parseRedshiftBoolean(data)
+          case DateType => parseDate(data)
+          case DoubleType => data.toDouble
+          case FloatType => data.toFloat
+          case IntegerType => data.toInt
+          case LongType => data.toLong
+          case ShortType => data.toShort
+          case StringType => data
+          case TimestampType => parseTimestamp(data)
+          case _ => data
+        }
+    }
+
+    Row(converted: _*)
+  }
+
+  /**
+   * Return a function that will convert arrays of strings conforming to
+   * the given schema to Row instances
+   */
+  def rowConverter(schema: StructType) = convertRow(schema, _: Array[String])
+
+
+  /**
+   * Convert schema representation of Dates to Timestamps, as spark-avro only works with Timestamps
+   */
+  def datesToTimestamps(sqlContext: SQLContext, df: DataFrame): DataFrame = {
+    val schema = StructType(
+      df.schema map {
+        case StructField(name, DateType, nullable, meta) => StructField(name, TimestampType, nullable, meta)
+        case other => other
+      })
+
+    sqlContext.createDataFrame(df.rdd, schema)
```

Comment:
We should probably use a cast here instead of just changing the schema (in case internal details about the difference between date and timestamp ever change).